index.coffee

```diff
@@ -26,21 +26,30 @@ config = {
 invoke = (handler, opts, cb) ->
 	fn = (callback) -> handler(opts, callback)
 
-	return if cb then fn(cb) else config.Promise (resolve, reject) ->
-		resolver = (err, args...) ->
-			resolver.invoked = true
-			if err?
-				reject(err)
-			else
-				resolve(if args.length > 1 then args else args[0])
-		resolver.invoked = false
-
-		try
-			maybePromise = fn(resolver)
-			if typeof maybePromise isnt 'undefined' and not resolver.invoked
-				resolver(null, maybePromise)
-		catch e
-			resolver(e)
+	if cb
+		maybePromise = fn(cb)
+		# Handle the case that request-promise might be enabled whilst using callbacks
+		# and therefore any errors will cause an unhandled rejection warning.
```

Comment:
This is the part of request-promise that handles setting up the response: https://github.com/request/promise-core/blob/master/lib/plumbing.js#L68-L133
And specifically https://github.com/request/promise-core/blob/master/lib/plumbing.js#L74-L81 is the part that calls the passed in callback but it's run separately from the promise that gets returned, so the returned promise never has a catch attached and so causes the unhandled promise rejection.  This is a problem specifically because requestmock doesn't interact with the special request loading the request promise uses in https://github.com/request/request-promise/blob/master/lib/rp.js#L9-L15 which means that the promise returning functionality is added to the global request rather than a copy, so places that only expect the callback interface will trigger the unhandled rejection warning

Reply:
Gotcha, thanks 👍